File: 4th Semester/Systems for Design and Implementation/C#/Homework2/OnwardTravel/Repository/Database/ReservationDbRepository.cs
﻿using log4net;
using log4net.Util;
using OnwardTravel.Models;
using OnwardTravel.Utils;
using System.Data;

namespace OnwardTravel.Repository.Database
{
    internal class ReservationDbRepository
        : IReservationRepository
    {
        private readonly DbUtils DbUtils;
        private static readonly ILog Logger = LogManager.GetLogger("ReservationDbRepository");

        public ReservationDbRepository(Properties props)
        {
            DbUtils = new(props);
        }

        public Guid? Save(Reservation entity)
        {
            Logger.InfoFormat("Saving the reservation: {0}", entity);
            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "INSERT INTO reservations(client_id, ride_id) VALUES (@client_id, @ride_id)";
                var paramClientID = comm.CreateParameter();
                paramClientID.ParameterName = "@client_id";
                paramClientID.Value = entity.Client.ID;
                comm.Parameters.Add(paramClientID);

                var paramRideID = comm.CreateParameter();
                paramRideID.ParameterName = "@ride_id";
                paramRideID.Value = entity.Ride.ID;
                comm.Parameters.Add(paramRideID);

                if (comm.ExecuteNonQuery() == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
            }
            using (var comm1 = DbUtils.GetConnection().CreateCommand())
            {
                comm1.CommandText = "SELECT reservation_id FROM Reservations WHERE client_id = @client_id AND ride_id = @ride_id";
                var paramClientID = comm1.CreateParameter();
                paramClientID.ParameterName = "@client_id";
                paramClientID.Value = entity.Client.ID;
                comm1.Parameters.Add(paramClientID);

                var paramRideID = comm1.CreateParameter();
                paramRideID.ParameterName = "@ride_id";
                paramRideID.Value = entity.Ride.ID;
                comm1.Parameters.Add(paramRideID);

                var result = DbUtils.FindQueryResults(comm1, DbUtils.ExtractGuid);
                if (result.Count == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
                Guid reservationID = result[0];

                using (var comm2 = DbUtils.GetConnection().CreateCommand())
                {
                    comm2.CommandText = "INSERT INTO reserved_seats(reservation_id, seat_number) VALUES (@reservation_id, @seat_number)";
                    var paramReservationID = comm2.CreateParameter();
                    paramReservationID.ParameterName = "@reservation_id";
                    paramReservationID.Value = reservationID;
                    comm2.Parameters.Add(paramReservationID);

                    foreach (var seat in entity.Seats)
                    {
                        var paramSeatID = comm2.CreateParameter();
                        paramSeatID.ParameterName = "@seat_number";
                        paramSeatID.Value = seat.ID;
                        comm2.Parameters.Add(paramSeatID);

                        comm2.ExecuteNonQuery();
                        comm2.Parameters.Remove(paramSeatID);
                    }
                }

                Logger.InfoExt(reservationID);
                return reservationID;
            }
        }

        public Reservation? Find(Guid identifier)
        {
            throw new NotImplementedException();
        }

        public Reservation? Update(Reservation entity)
        {
            throw new NotImplementedException();
        }

        public Reservation? Delete(Guid identifier)
        {
            throw new NotImplementedException();
        }

        public IList<Reservation> FindReservationsOnRide(Ride ride)
        {
            Logger.InfoFormat("Finding the reservations for the ride: {0}", ride);

            IList<Reservation> reservations = new List<Reservation>();
            using (var comm1 = DbUtils.GetConnection().CreateCommand())
            {
                comm1.CommandText = "SELECT reservation_id, client_id FROM reservations WHERE ride_id = @ride_id";
                var paramRideID = comm1.CreateParameter();
                paramRideID.ParameterName = "@ride_id";
                paramRideID.Value = ride.ID;
                comm1.Parameters.Add(paramRideID);

                using (var dataR1 = comm1.ExecuteReader())
                {
                    while (dataR1.Read())
                    {
                        Guid reservationID = dataR1.GetGuid(0);
                        Guid clientID = dataR1.GetGuid(1);
                        Client? client = null;
                        IList<ReservedSeat> seats = new List<ReservedSeat>();

                        using (var comm2 = DbUtils.GetConnection().CreateCommand())
                        {
                            comm2.CommandText = "SELECT * FROM clients WHERE client_id = @client_id";
                            var paramClientID = comm2.CreateParameter();
                            paramClientID.ParameterName = "@client_id";
                            paramClientID.Value = clientID;
                            comm2.Parameters.Add(paramClientID);

                            using (var dataR2 = comm2.ExecuteReader())
                            {
                                if (dataR2.Read())
                                {
                                    client = ClientDbRepository.ExtractClient(dataR2);
                                }
                            }
                        }

                        if (client == null)
                        {
                            continue;
                        }

                        using (var comm3 = DbUtils.GetConnection().CreateCommand())
                        {
                            comm3.CommandText = "SELECT reserved_seat_id, seat_number FROM reserved_seats WHERE reservation_id = @reservation_id";
                            var paramReservationID = comm3.CreateParameter();
                            paramReservationID.ParameterName = "@reservation_id";
                            paramReservationID.Value = reservationID;
                            comm3.Parameters.Add(paramReservationID);

                            using (var dataR3 = comm3.ExecuteReader())
                            {
                                while (dataR3.Read())
                                {
                                    Guid guid = dataR3.GetGuid(0);
                                    int seatNumber = dataR3.GetInt32(1);
                                    seats.Add(new ReservedSeat(guid, seatNumber));
                                }
                            }
                        }
                        reservations.Add(new Reservation(reservationID, client, ride, seats));
                    }
                }
            }

            Logger.InfoExt(reservations);
            return reservations;
        }
    }
}

File: 4th Semester/Systems for Design and Implementation/Java/Homework6/OnwardPersistence/src/main/java/repository/database/ReservationDbRepository.java
package repository.database;

import models.Client;
import models.Reservation;
import models.ReservedSeat;
import models.Ride;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.ReservationRepository;
import repository.exceptions.DataChangeException;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;
import java.util.UUID;

public class ReservationDbRepository implements ReservationRepository {
    private final DbUtils dbUtils;
    private static final Logger logger = LogManager.getLogger(ReservationDbRepository.class);

    public ReservationDbRepository(Properties props) {
        logger.traceEntry("Initializing UserDbRepository with properties: {}", props);
        dbUtils = new DbUtils(props);
    }

    @Override
    public UUID save(Reservation entity) throws DataChangeException {
        logger.traceEntry("Saving the reservation: {}", entity);
        UUID reservationID;
        try (PreparedStatement preparedStatement0 = dbUtils.getConnection().prepareStatement(
                "SELECT reservation_id FROM reservations WHERE client_id = ? AND ride_id = ?")) {
            preparedStatement0.setObject(1, entity.getClient().getID(), Types.OTHER);
            preparedStatement0.setObject(2, entity.getRide().getID(), Types.OTHER);
            var result = DbUtils.findQueryResult(preparedStatement0, DbUtils::extractUUID);
            reservationID = result.isEmpty() ? null : result.get(0);
        } catch (SQLException ex) {
            logger.error(ex);
            throw new DataChangeException(ex);
        }

        if (reservationID != null) {
            try (PreparedStatement preparedStatement = dbUtils.getConnection().prepareStatement(
                    "INSERT INTO reservations(client_id, ride_id) VALUES (?, ?)")) {
                preparedStatement.setObject(1, entity.getClient().getID(), Types.OTHER);
                preparedStatement.setObject(2, entity.getRide().getID(), Types.OTHER);

                if (preparedStatement.executeUpdate() == 0) {
                    logger.traceExit(null);
                    return null;
                }
            } catch (SQLException ex) {
                logger.error(ex);
                throw new DataChangeException(ex);
            }
            try (PreparedStatement preparedStatement1 = dbUtils.getConnection().prepareStatement(
                    "SELECT reservation_id FROM reservations WHERE client_id = ? AND ride_id = ?")) {
                preparedStatement1.setObject(1, entity.getClient().getID(), Types.OTHER);
                preparedStatement1.setObject(2, entity.getRide().getID(), Types.OTHER);
                var result = DbUtils.findQueryResult(preparedStatement1, DbUtils::extractUUID);
                reservationID = result.isEmpty() ? null : result.get(0);
                if (reservationID == null) {
                    logger.traceExit(null);
                    return null;
                }
            } catch (SQLException ex) {
                logger.error(ex);
                throw new DataChangeException(ex);
            }
        }
        try (PreparedStatement preparedStatement2 = dbUtils.getConnection().prepareStatement(
                "INSERT INTO reserved_seats(reservation_id, seat_number) VALUES (?, ?)")) {
            preparedStatement2.setObject(1, reservationID, Types.OTHER);

            for (var seat : entity.getSeats()) {
                preparedStatement2.setInt(2, seat.getSeatNumber());
                preparedStatement2.executeUpdate();
            }
            return reservationID;
        } catch (SQLException ex) {
            logger.error(ex);
            throw new DataChangeException(ex);
        }
    }

    @Override
    public Reservation find(UUID identifier) {
        return null;
    }

    @Override
    public Reservation update(Reservation entity) {
        return null;
    }

    @Override
    public Reservation delete(UUID identifier) {
        return null;
    }

    @Override
    public List<Reservation> findReservationsByRide(Ride ride) {
        logger.traceEntry("Finding the reservations for the ride: {}", ride);

        List<Reservation> reservations = new ArrayList<>();
        try (PreparedStatement preparedStatement1 = dbUtils.getConnection().prepareStatement(
                "SELECT reservation_id, client_id FROM reservations WHERE ride_id = ?")) {
            preparedStatement1.setObject(1, ride.getID(), Types.OTHER);
            try (ResultSet resultSet1 = preparedStatement1.executeQuery()) {
                while (resultSet1.next()) {
                    UUID reservationID = resultSet1.getObject("reservation_id", UUID.class);
                    UUID clientID = resultSet1.getObject("client_id", UUID.class);

                    Client client;
                    try (PreparedStatement preparedStatement2 = dbUtils.getConnection().prepareStatement(
                            "SELECT * FROM clients WHERE client_id = ?"
                    )) {
                        preparedStatement2.setObject(1, clientID);
                        var result = DbUtils.findQueryResult(preparedStatement2, ClientDbRepository::extractClient);
                        client = result.isEmpty() ? null : result.get(0);
                    }

                    if (client == null) {
                        continue;
                    }
                    List<ReservedSeat> seats = new ArrayList<>();

                    try (PreparedStatement preparedStatement3 = dbUtils.getConnection().prepareStatement(
                            "SELECT reserved_seat_id, seat_number FROM reserved_seats WHERE reservation_id = ?")) {
                        preparedStatement3.setObject(1, reservationID);
                        try (ResultSet resultSet3 = preparedStatement3.executeQuery()) {
                            while (resultSet3.next()) {
                                UUID reservationSeatID = resultSet3.getObject("reserved_seat_id", UUID.class);
                                int seatNumber = resultSet3.getInt("seat_number");

                                seats.add(new ReservedSeat(reservationSeatID, seatNumber));
                            }
                        }
                    }
                    reservations.add(new Reservation(client, ride, seats));
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            throw new RuntimeException(ex);
        }

        return logger.traceExit(reservations);
    }
}
